genesis/storage.go

```diff
@@ -0,0 +1,180 @@
+	if state.IsNil(st) || err != nil {
+		return nil, err
+	}
+	return st.Copy(), nil
+}
+
+func stateInternal() (state.BeaconState, error) {
+	gd := getPkgVar()
+	if !gd.initialized {
+		// If the state is not explicitly initialized, try to load embedded states if available.
+		name := params.BeaconConfig().ConfigName
+		if ed, ok := embeddedGenesisData[name]; ok {
+			return ed.embeddedState()
+		}
+		return nil, ErrGenesisStateNotInitialized
+	}
+	if !state.IsNil(gd.State) {
+		return gd.State, nil
+	}
+	if gd.embeddedState != nil {
+		st, err := gd.embeddedState()
+		if err != nil {
+			return nil, errors.Wrap(err, "load embedded genesis state")
+		}
+		if !state.IsNil(st) {
+			gd.State = st
+			return gd.State, nil
+		}
+	}
+	return loadState()
+}
+
+// Store is an exported method that allows another package to set the genesis data value and persist it to disk.
+// It is exported to be used by implementations of the Provider interface.
+func Store(d GenesisData) error {
+	if err := ensureWritable(d.FileDir); err != nil {
+		return err
+	}
+	if err := persist(d); err != nil {
+		return errors.Wrap(err, "persist genesis data")
+	}
+	setPkgVar(d, true)
+	return nil
+}
+
+type fnamePart int
+
+const (
+	genesisPart fnamePart = 0
+	timePart    fnamePart = 1
+	gvrPart     fnamePart = 2
+)
+
+// data is a private package level variable that holds the genesis data.
+// Other packages interact with it via wrapper functions like Set() and State().
+var data GenesisData
+var stateMu sync.Mutex
+
+// GenesisData bundles all the package level data. It is exported to allow implementations of the Provider interface to set genesis data.
+type GenesisData struct {
+	ValidatorsRoot [32]byte
+	Time           time.Time
+	FileDir        string
+	State          state.BeaconState
+	embeddedBytes  func() ([]byte, error)
+	embeddedState  func() (state.BeaconState, error)
+	initialized    bool
+}
+
+func (d GenesisData) filePath() string {
+	parts := [3]string{}
+	parts[genesisPart] = "genesis"
+	parts[timePart] = strconv.FormatInt(d.Time.Unix(), 10)
+	parts[gvrPart] = hexutil.Encode(d.ValidatorsRoot[:])
+	return path.Join(d.FileDir, strings.Join(parts[:], "-")+".ssz")
+}
+
+func persist(d GenesisData) error {
+	if state.IsNil(d.State) {
+		return ErrGenesisStateNotInitialized
+	}
+	if d.FileDir == "" {
+		return ErrFilePathUnset
+	}
+	fpath := d.filePath()
+	sb, err := d.State.MarshalSSZ()
+	if err != nil {
+		return errors.Wrap(err, "marshal ssz")
+	}
+	if err := file.WriteFile(fpath, sb); err != nil {
+		return fmt.Errorf("error writing genesis state to %s: %w", fpath, err)
+	}
+	log.WithField("filePath", fpath).Info("Genesis state written to disk.")
+	return nil
+}
+
+func getPkgVar() GenesisData {
+	stateMu.Lock()
+	defer stateMu.Unlock()
+	return data
+}
+
+func setPkgVar(d GenesisData, initialized bool) {
+	stateMu.Lock()
+	defer stateMu.Unlock()
+	d.initialized = initialized
+	data = d
+}
+
+func loadState() (state.BeaconState, error) {
+	stateMu.Lock()
```

Comment:
do we need to lock here? what about moving it to line 149 after reading from the file?

Reply:
Yeah this was kind of lazy, just locking around this whole func. Although, we do only want to read once, but there should ideally be a lock around the nil check also for that to work, otherwise we'll still read twice. I'll clean this up.